streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -232,40 +239,54 @@ private void processEarly(final K key, final V value, final long timestamp, fina
                     final long startTime = next.key.window().start();
                     final long endTime = startTime + windows.timeDifferenceMs();
 
-                    if (endTime == windows.timeDifferenceMs()) {
+                    if (startTime == 0) {
                         combinedWindow = next;
-                    } else if (endTime > timestamp && startTime <= timestamp) {
+                    } else if (endTime >= timestamp && startTime <= timestamp) {
                         rightWinAgg = next.value;
                         putAndForward(next.key.window(), next.value, key, value, closeTime, timestamp);
-                    } else {
+                    } else if (startTime == timestamp + 1) {
                         rightWinAlreadyCreated = true;
                     }
                 }
             }
 
+            // if there wasn't a right window agg found and we need a right window for our new record,
+            // the current aggregate in the combined window iwll go in the new record's right window
```

Comment:
`iwll` 🙂

Reply:
Also, cool,  I think I understand the concept here but some of the details are a bit fuzzy. Basically if we don't find a right window agg that means we didn't find any windows (besides the combined window), which in turn means that there can only be a single record in the combined window (otherwise you'd get a right window for the earlier record). 
So we need to use the combined window agg for the current record's right window.  But we should only do that if the one record is actually after the current record, right? I think you actually do implicitly check that is the case below but it's pretty subtle: basically in `previousRightWindowPossible` you would return false if `rightWindowStart > currentRecordTimestamp`. But we can check that right here and make it explicit, so that `rightWinAgg` only ever means the aggregate that we will actually put in the current record's right window. Then we can also clean up `previousRightWindowPossible`

Reply:
There can be more than one record in the combined window, but if we don't find any right window agg then there are no records to the left of this record (in time) AND to the right of this record (could be one or the other). There could be multiple to the right though.

It took me like 5 minutes to figure this out, but actually the call to `previousRightWindowPossible` is when we're creating the right window for the previous record, not for the current record. This call doesn't involved `rightWindowAgg` at all, since the check `rightWindowStart > currentRecordTimestamp` wouldn't help us figure out if `previous record > current record`.

The check for this is actually a little further down (line 277) where we do `rightWindowIsNotEmpty`, and there it confirms that the values in rightWinAgg are > timestamp.

Reply:
Couldn't there still be a record to the left? Like we could have a record at 5 and at 50 and nothing else, then all we would have so far is the combined window and one at [40, 50],  but `rightWindowAgg` would be null. So that's why we need to check that the `combinedWindow.maxTimestamp > timestamp` (and if not then we should leave `rightWinAgg` as null). Looks like this is what you're doing, so not suggesting any changes, just trying to make sure I have this right.